### PR TITLE
Fix GCNN initializers

### PR DESCRIPTION
**Pull request by attila-i-szabo**
Currently, the initializers used for GCNNs are a bit of a mess. We define several different initializers (some of which are manifestly incorrect, e.g. a `jax.nn.lecun_normal` that refers to the wrong axes), all of which are trumped in the `GCNN` classes by a custom variance-scaled normal distribution, ~which is however a factor of sqrt(2) too large for complex numbers.~ this only works because `jax.random.normal` is now defined to halve the variance of real/imaginary parts for complex dtypes.

This PR clears this up. 
* The one remaining initialiser in `nk.nn` is dropped for it is internal and now available in all versions of JAX that we support. This leaves only deprecations in the `initializers.py` file.
* A single `default_gcnn_initializer` is defined in the same file as `DenseSymm` and `DenseEquivariant`, which is also exported into `GCNN`: this is the plain JAX `lecun_normal` instantiated with the correct axes, the one we ought to use with SELU.

Variance scaling for masks is also fixed by defining a `scaled_mask` whose norm is always the same as a mask full of 1s.

Possible TODOs:
* Check the initialiser of `RBMSymm`. I have no intuition as to whether `lecun_normal` is appropriate for that too, so please weigh in; we can probably do better than the current `normal(0.1)`
* More generally, define similar initialisers for non-GCNN layers like `Dense`

### CODE REVIEW

**Comment by PhilipVinc** [comment]
is this equivalent to the old?

**Comment by attila-i-szabo** [comment]
No, the point is that the initializers are changed to do variance scaling correctly

**Comment by PhilipVinc** [comment]
Last time I tried to do that in https://github.com/netket/netket/pull/863 I found out after a little benchmarking that the scale invariant initialisation was working great with SGD and quite poorly with SR that then needed more tuning, so we did not merge it.

Can you try to run a bunch of benchmark simulations both with SGD and SR for different sizes before and after this PR to see how things change?

**Comment by attila-i-szabo** [comment]
Thanks for this. I know nothing about RBM, so if you say it wasn't working great, I won't push to change it in any way.

As I said, deeper GCNNs have been scale-invariant already (the `unit_normal_scaling` that I've deleted is equivalent to `lecun_normal` without truncation), and it has to be if it is not to blow up. I've tried this code in production (on a 4-layer network) and it seems to work just as well as the previous one. SR usually gets slow later in the training when it's near the ground state where the S matrix is the worst-conditioned.

**Comment by PhilipVinc** [comment]
I don't want to stop you from doing this! If this works, i'm all in for merging it. I'm just asking for a bunch of plots like https://github.com/netket/netket/pull/863#issuecomment-899335451 but both with and without sr

File: netket/nn/symmetric_linear.py
Comment (chrisrothUT) [inline netket/nn/symmetric_linear.py:34]: I'm pretty sure `in_axis = (1,2)` here. And don't we have to specify `mode="fan_in"`?

**Review by chrisrothUT — CHANGES_REQUESTED** [review]

File: netket/nn/symmetric_linear.py
Comment (attila-i-szabo) [reply netket/nn/symmetric_linear.py:34]: Yeah, that's confusing, `in_axis` is the axes that **only** appear in the input but not the output, so axis 2 is neither. And `lecun_normal` implies `fan_in`

Comment (attila-i-szabo) [reply netket/nn/symmetric_linear.py:34]: See the code [here](https://jax.readthedocs.io/en/latest/_modules/jax/_src/nn/initializers.html)

Comment (chrisrothUT) [reply netket/nn/symmetric_linear.py:34]: Ahh I see, yes the code as written should work.

**Review by chrisrothUT — APPROVED** [review]

**Comment by attila-i-szabo** [comment]
OK. Benchmarks. As predicted, the performance of GCNNs doesn't noticeably change (I think any difference in these curves is within the range of outcomes for one initializer).

I used two Hamiltonians, both on the square lattice
* NN antiferromagnet with Marshall sign rule → use real-parameter network
* J1-J2 antiferromagnetic (J2=J1/2) → use complex-parameter network

In all figures, orange is the new initialiser, blue is the old one.

SR, NN model, 4×4 lattice, 6 layers, 4 features
image

SR, J1-J2, 4×4 lattice, 6 layers, 4 features
image

SR, J1-J2, 6×6 lattice, 4 layers, 4 features
image

SGD, NN model, 6×6 lattice, 2 layers, 4 features
image

SGD was notoriously hard to stabilise with either initializer, it didn't work at all for the J1-J2 model, nor for deeper networks with the NN model.

**Review by femtobit — APPROVED** [review]
Looks good to me, I have no objections. I'll leave the decision to merge to @PhilipVinc.

**Comment by chrisrothUT** [comment]
There's also a minor bug in the GCNN models which is that this line

`x = jnp.expand_dims(x, -2) ` # add a feature dimension

should really be  

```
if x.ndim < 3:
  x = jnp.expand_dims(x, -2)
```

so that we only expand the dimension if we don't have a onehot representation. Maybe we could sneak that minor change into this PR

**Comment by attila-i-szabo** [comment]
Good catch. There's an equivalent line in `RBMSymm`, I changed that too.

We happy, @PhilipVinc'ent?

**Comment by PhilipVinc** [comment]
yes I am happy though can you add a changeling notice for both the bug fix and the PR itself? (While you are at it, since it's the first commit, can you change the version 3.4 banner to 3.3.1 in the changelog and remove all sections other than  bug fix?

**Review by PhilipVinc — APPROVED** [review]
I'll merge when a changelog notice is added

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:33]: I just have an extra comment: This file is called `symmetric_linear` and contains equivariant layers. GCNN is a NN architecture built on top of equivariant layers that, in principle, we should be able to use alone without the gcnn.

It's therefore a bit weird for me to find here `default_gcnn_initializer` defined. I'd much more like an initialiser named `default_gcnn_initializer`. Probably a better name could be something like `default_equivariant_initializer` or `default_symmetry_initializer`? Maybe @jwnys has a better idea for a name.

File: netket/models/equivariant.py
Comment (PhilipVinc) [inline netket/models/equivariant.py:36]: Ah, I just noticed now, but...
in principle all things you are importing here from `symmetric_linear` are 'implementation details. You should only access `netket.nn.DenseSymm` and `netket.nn.DenseEquivariant`.

While this is intra-netket and using internals is fine, I'd like to avoid it if possible.

For what regards this particular PR, I'd be fine with not importing the default initialiser and just redefining it here + a comment below.

Then, it would be great if in a follow up PR you could just use the standard API `DenseSymm(mode="FFT")` or `DenseEquivariant(mode=irrep)` instead of using the type directly.

This (1) tests the public API, so it's good, (2) makes us use our public API and (3) disentangles the internals of the two modules.

Moreover, maybe by doing this we can get rid of having 4 different GCNN classes?

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:34]: By the way, is this supported in all jax versions that we support? I remember talking about raising the jax version with @attila-i-szabo ?

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:79]: The mask must be an hashablearray, right? `jnp.linalg.norm(self.mask)` works because it's doing essentially `jnp.linalg.norm(jnp.array(np.asarray(self.mask)))`. 
But you are doing

```python
self.mask / jnp.linalg.norm(Self.mask)
```
which is converted to
```
jnp.array(self.mask) / jnp.linalg.norm(jnp.array(np.asarray(self.mask)))
```
which captures the mask twice.

I think a better implementation would be to capture only the rescaled mask once:

```suggestion
        if self.mask is not None:
            self.scaled_mask = jnp.asarray(
                self.mask / np.linalg.norm(self.mask) * self.n_sites ** 0.5)
```
(probably you need to import numpy and reformat too)

Comment (PhilipVinc) [reply netket/nn/symmetric_linear.py:79]: Also can we assert that the bask is boolean and not integers or other things? 
Or do we need in some cases the mask to be a finite number?

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:173]: as above

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:411]: as above. maybe you can make a function ?

File: netket/nn/symmetric_linear.py
Comment (PhilipVinc) [inline netket/nn/symmetric_linear.py:561]: .

File: test/models/test_nn.py
Comment (PhilipVinc) [inline test/models/test_nn.py:190]: do we test that the mask actually masks?

**Review by PhilipVinc — CHANGES_REQUESTED** [review]

File: netket/nn/symmetric_linear.py
Comment (attila-i-szabo) [reply netket/nn/symmetric_linear.py:34]: Yeah, this is fine for both `float` and `complex`. It would've been a problem if we kept 2D kernels in `DenseSymm` because that's only supported from 2.26

File: test/models/test_nn.py
Comment (attila-i-szabo) [reply test/models/test_nn.py:190]: wdym?

File: netket/nn/symmetric_linear.py
Comment (attila-i-szabo) [reply netket/nn/symmetric_linear.py:79]: I've never used this feature so no strong feelings, but it might be useful to control the initialisation through this mask. The scaling logic is happy with that too.

File: netket/models/equivariant.py
Comment (attila-i-szabo) [reply netket/models/equivariant.py:36]: `DenseSymm()` and `DenseEquivariant()` take things that don't play nicely at all with jitting, so if you put these functions in `setup()` of another module, their construction gets very slow. I seem to recall we tried this at some point (but @chrisrothUT  might remember better) and decided against it when we saw the performance.

File: netket/nn/symmetric_linear.py
Comment (attila-i-szabo) [reply netket/nn/symmetric_linear.py]: I generally don't like restricting inputs unless they break stuff/lead to something very absurd. This is neither.

**Review by PhilipVinc — APPROVED** [review]
Thanks for addressing all my comments @attila-i-szabo .

I think one test that is missing and if you have some spare minutes in the future would be one checking that the mask is actually masking (in the sense that it affects the output in the correct way).

Anyhow, this is great and I'll merge it.

**Comment by PhilipVinc** [comment]
Another thing that might be worth investigating briefly before releasing this is if lecun_normal has the right variance wrt the default activation (reim_selu)

**Comment by attila-i-szabo** [comment]
> I think one test that is missing and if you have some spare minutes in the future would be one checking that the mask is actually masking (in the sense that it affects the output in the correct way).

Fair enough, the masking logic is hardly tested in general, I'll try to tend to this at some point.

> Another thing that might be worth investigating briefly before releasing this is if lecun_normal has the right variance wrt the default activation (reim_selu)

I did check this (in fact that's why I started this PR). `lecun_normal` ensures that the expected |x|^2 remains 1 for both real and complex in all layers. SELU normally wants real/imag. parts with variance 1; we added factors of sqrt(2) to `reim_selu` with complex inputs to correct for this.

**Pull request merged**
